### PR TITLE
fix(key): change to string only

### PR DESCRIPTION
**Pull request by StrangeGirlMurph**
I probably missed a few things here and there. I didn't run any tests. Please do that if necessary. Also please rename this pr :)
If I messed something up I will probably find it pretty soon since we're actively using enmap a lot on a project.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1005,15 +999,31 @@ class Enmap extends Map {
...
-   * INTERNAL method to verify the type of a key or property
+  /** 
+   * INTERNAL METHOD. Checks if the key is valid.
+   * Will THROW AN ERROR if the key is nil or not a string.
+   * @param {string|Array<string>} keys The key(s) to check.
+   */
+  #checkKeyOrArrayOfKeys(keys) {
+    if (!(typeof keys === 'string' || (Array.isArray(keys) && keys.every(e => typeof e === "string")))) {
+      throw new Err(
+        `Enmap requires keys to be a string. Provided: ${
```
Comment (dan-online) [inline src/index.js:1010]: Or array of strings

File: src/index.js [hunk 2]
```diff
@@ -467,6 +457,7 @@ class Enmap extends Map {
    * @returns {Enmap} The enmap.
    */
   push(key, val, path = null, allowDupes = false) {
+    this.#checkKeyOrArrayOfKeys(key);
     const data = this.get(key);
     this.#check(key, 'Array', path);
```
Comment (dan-online) [inline src/index.js:462]: Could we move the checkKeyOrArrayOfKeys to the check function?

File: src/index.js [hunk 3]
```diff
@@ -717,7 +711,7 @@ class Enmap extends Map {
       this.set(key, data);
     } else {
       super.delete(key);
-      if (typeof this.changedCB === 'function') {
+      if (typeof this.changedCB === 'function') {0
```
Comment (dan-online) [inline src/index.js:714]: Why the 0?

File: src/index.js [hunk 4]
```diff
@@ -1005,15 +999,31 @@ class Enmap extends Map {
...
+  /**
+   * INTERNAL METHOD to verify a property
    * Will THROW AN ERROR on wrong type, to simplify code.
    * @param {string} key Required. The key of the element to check
    * @param {string} type Required. The javascript constructor to check
    * @param {string} path Optional. The dotProp path to the property in the object enmap.
    */
   #check(key, type, path = null) {
-    key = key.toString();
+    this.#checkKeyOrArrayOfKeys(key);
```
Comment (dan-online) [inline src/index.js:1026]: Oh you did, then we can remove redundant cases

File: typings/index.d.ts [hunk 5]
```diff
@@ -353,7 +369,7 @@ declare module 'enmap' {
      * points.inc("numberInObject", "sub.anInt"); // {sub: { anInt: 6 }}
      * @returns The enmap.
      */
-    public inc(key: K, path?: string): this;
+    public inc(key: string, path?: string): this;
```
Comment (dan-online) [inline typings/index.d.ts:372]: What happened to string[]?

File: typings/index.d.ts [hunk 6]
```diff
@@ -433,47 +459,57 @@ declare module 'enmap' {
...
+    public deleteProp(key: string, path: string): void;
 
     /**
      * Deletes everything from the enmap. If persistent, clears the database of all its data for this table.
      */
     public deleteAll(): void;
 
+    /**
+     * Completely destroys the entire enmap. This deletes the database tables entirely.
+     * It will not affect other enmap data in the same database, however.
```
Comment (dan-online) [inline typings/index.d.ts:479]: Why the comma?

File: typings/index.d.ts [hunk 7]
```diff
@@ -433,47 +459,57 @@ declare module 'enmap' {
...
+     * THIS ACTION WILL DESTROY YOUR DATA AND CANNOT BE UNDONE.
+     * @returns {null}
+     */
+    public destroy(): null;
+
+    /**
+     * Deletes everything from the enmap. If persistent, clears the database of all its data for this table.
```
Comment (dan-online) [inline typings/index.d.ts:486]: Maybe also the caps warning of cannot be undone

**Review by dan-online — CHANGES_REQUESTED** [review]
Few changes, also I'll wait until #103 merges before another review

File: typings/index.d.ts (hunk 6 above)
Comment (WilsontheWolf) [reply typings/index.d.ts:479]: That's how grammar works. However is a connective that has a comma before and after it

File: src/index.js [hunk 8]
```diff
@@ -1005,15 +999,31 @@ class Enmap extends Map {
...
 
-  /*
-   * INTERNAL method to verify the type of a key or property
+  /** 
+   * INTERNAL METHOD. Checks if the key is valid.
+   * Will THROW AN ERROR if the key is nil or not a string.
+   * @param {string|Array<string>} keys The key(s) to check.
+   */
+  #checkKeyOrArrayOfKeys(keys) {
+    if (!(typeof keys === 'string' || (Array.isArray(keys) && keys.every(e => typeof e === "string")))) {
```
Comment (WilsontheWolf) [inline src/index.js:1008]: I don't believe a decent chunk of the methods using this support an array of keys.

File: src/index.js [hunk 9]
```diff
@@ -1098,20 +1108,18 @@ class Enmap extends Map {
   }
 
   /**
-   * Internal method used to validate persistent enmap names (valid Windows filenames)
+   * INTERNAL METHOD used to validate persistent enmap names (valid Windows filenames)
```
Comment (WilsontheWolf) [inline src/index.js:1111]: If we're being consistent 
```suggestion
   * INTERNAL METHOD. Validates persistent enmap names (valid Windows filenames)
```

**Review by WilsontheWolf — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 3 above)
Comment (StrangeGirlMurph) [reply src/index.js:714]: typo

File: typings/index.d.ts (hunk 5 above)
Comment (StrangeGirlMurph) [reply typings/index.d.ts:372]: string[] was never a part of the function and is not supported by it.

File: typings/index.d.ts (hunk 6 above)
Comment (StrangeGirlMurph) [reply typings/index.d.ts:479]: I dunno. I just copied the description from the code. I am really not the best person to ask about grammar stuff. Just edit it to your likings :)

File: src/index.js (hunk 8 above)
Comment (StrangeGirlMurph) [reply src/index.js:1008]: Yes but the two extra checks shouldn't hurt.

File: src/index.js (hunk 4 above)
Comment (StrangeGirlMurph) [reply src/index.js:1026]: I am a bit unsure... Some public functions use check() some don't. I feel like in general is there a lot of redundancy in the code and a consistency missing.

File: src/index.js (hunk 8 above)
Comment (WilsontheWolf) [reply src/index.js:1008]: Well it would cause it you pass an array of strings it now will accept it

Comment (StrangeGirlMurph) [reply src/index.js:1008]: Oh I am dumb. You're right... That's a problem... One moment

File: src/index.js [hunk 10]
```diff
@@ -1000,21 +1000,66 @@ class Enmap extends Map {
...
+   * INTERNAL METHOD. Checks if the key or array of keys is valid.
+   * Will THROW AN ERROR if the key is neither a string nor an array of strings.
    * @param {string|Array<string>} keys The key(s) to check.
    */
   #checkKeyOrArrayOfKeys(keys) {
-    if (!(typeof keys === 'string' || (Array.isArray(keys) && keys.every(e => typeof e === "string")))) {
+    if (Array.isArray(keys)) {
+      this.#checkArrayOfKeys(keys);
+    } else {
+      this.#checkKey(keys);
```
Comment (WilsontheWolf) [inline src/index.js:1011]: Doing this this way means the error won't say it can accept an array of a non-array is passed

**Review by WilsontheWolf — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 10 above)
Comment (StrangeGirlMurph) [reply src/index.js:1011]: uh?

File: src/index.js (hunk 4 above)
Comment (dan-online) [reply src/index.js:1071]: Perhaps specify check() in each method call

File: typings/index.d.ts (hunk 6 above)
Comment (dan-online) [reply typings/index.d.ts:481]: > That's how grammar works. However is a connective that has a comma before and after it

Ah you're right, it's an interrupter in this case.

File: src/index.js (hunk 4 above)
Comment (StrangeGirlMurph) [reply src/index.js:1071]: 🤷‍♀️ I dunno. It should work how it is right now. If you want to remove the redundant checks feel free to edit it yourself.

Comment (dan-online) [reply src/index.js:1071]: Well no, you made the pull request 😂, please remove the redundant checks

File: src/index.js (hunk 10 above)
Comment (WilsontheWolf) [reply src/index.js:1011]: For example, if I pass a number to a method that may take an array, it will only say it can take a string

Comment (StrangeGirlMurph) [reply src/index.js:1011]: Yeah for me that's the expected behavior. I think you usually know if a variable is an Array or not. And the error also outputs which type you passed. The message that for example number and string don't align is enough.

Comment (WilsontheWolf) [reply src/index.js:1011]: Ok

File: src/index.js (hunk 2 above)
Comment (StrangeGirlMurph) [reply src/index.js:462]: `this.#check` is rather for checking a property. I removed the `this.#checkKeyOrArrayOfKeys(key);` to get rid of the redundant cases

**Comment by StrangeGirlMurph** [comment]
@dan-online 103 done I merged pls rereview

File: typings/index.d.ts [hunk 11]
```diff
@@ -793,15 +876,15 @@ declare module 'enmap' {
      * @example
      * const newColl = someColl.clone();
      */
-    public clone(): Enmap<K, V>;
+    public clone(): Enmap<V>;
```
Comment (dan-online) [inline typings/index.d.ts:879]:
```suggestion
    public clone(): Enmap<V, SV>;
```

**Review by dan-online — COMMENTED** [review]
Just one question and a couple of proposed changes

File: typings/index.d.ts [hunk 12]
```diff
@@ -793,15 +876,15 @@ declare module 'enmap' {
...
 
     /**
      * Combines this Enmap with others into a new Enmap. None of the source Enmaps are modified.
      * @param enmaps Enmaps to merge
      * @example
      * const newColl = someColl.concat(someOtherColl, anotherColl, ohBoyAColl);
      */
-    public concat(...enmaps: Enmap<K, V>[]): Enmap<K, V>;
+    public concat(...enmaps: Enmap<V>[]): Enmap<V>;
```
[the remark this answers is not in the document]
Comment (StrangeGirlMurph) [reply typings/index.d.ts:887]: mhm thanks

File: src/index.js [hunk 13]
```diff
@@ -202,17 +202,8 @@ class Enmap extends Map {
...
-    if (isNil(key) || key.constructor.name !== 'String') {
-      throw new Err(
-        `Enmap requires keys to be a string. Provided: ${
-          isNil(key) ? 'nil' : key.constructor.name
-        }`,
-        'EnmapKeyTypeError',
-      );
-    }
-    key = key.toString();
+    this.#checkKey(key);
```
[the remark this answers is not in the document]
Comment (StrangeGirlMurph) [reply src/index.js:206]: Yes thanks

**Review by dan-online — APPROVED** [review]
LGTM

**Comment by eslachance** [comment]
PR was reverted and removed from 5.9.5, multiple issues were found with the code, including issue #109 as well as `undefined` key now throwing an error instead of returning undefined in get. This is a breaking change that should not occur in a minor patch change.

**Pull request merged**